docs/tutorial.md

```diff
@@ -0,0 +1,721 @@
+
+    ```swift
+    class ViewController: UIViewController {
+
+    @IBOutlet var nameLabel: UILabel!
+    @IBOutlet var signInButton: UIButton!
+    @IBOutlet var signOutButton: UIButton!
+    ```
+
+    And XCode has added two outlines to the `Main.storyboard` source.
+
+    ```xml
+    <connections>
+    <outlet property="nameLabel" destination="9eE-ZS-LU9" id="Ahv-Te-ZZo"/>
+    <outlet property="signInButton" destination="Lfp-KX-BDb" id="yef-Jj-uPt"/>
+    <outlet property="signOutButton" destination="k0H-mY-bo5" id="Y47-Za-lSc"/>
+    </connections>
+    ```
+
+    Your connectors will have their own `destination` and `id` values.
+
+
+### Edit the ViewController.swift file
+
+Now, you are ready to connect your applicaiton with your Blockstack Web
+Application. Normally, after building your Web application you would have
+registred it with Blockstack and the app would be available on the Web. This
+example skips this registroation step and uses an example application we've
+already created for you:
+
+`https://heuristic-brown-7a88f8.netlify.com/redirect.html`
+
+This web application already has a redirect in place for you. You'll reference
+this application in your mobile add for now. In XCode, do the following;
+
+
+1. Open the `ViewController.swift` file.
+2. Add an import both for `Blockstack` and for `SafariServices`.
+
+   ```
+	 import UIKit
+	 import Blockstack
+	 import SafariServices
+	 ```
+
+4. Just before the `didReceiveMemoryWarning` function a private `updateUI` function.
+
+   This function takes care of loading the user data from Blockstack.
+
+   ```
+   private func updateUI() {
+     DispatchQueue.main.async {
+    		if Blockstack.shared.isSignedIn() {
+    			// Read user profile data
+    			let retrievedUserData = Blockstack.shared.loadUserData()
+    			print(retrievedUserData?.profile?.name as Any)
+    			let name = retrievedUserData?.profile?.name ?? "Nameless Person"
+    			self.nameLabel.text = "Hello, \(name)"
+    			self.nameLabel.isHidden = false
+            self.signInButton.isHidden = true
+            self.signOutButton.isHidden = false
+    		} else {
+            self.signInButton.isHidden = false
+            self.signOutButton.isHidden = true
+            self.nameLabel.isHidden = true
+    		}
+    	}
+    }
+    ```
+
+5. Replace the content of the `viewDidLoad()` function so that it calls this private function.
+
+   ```
+	 override func viewDidLoad() {
+		self.updateUI()
+   }
+	 ```
+
+6. Add a `handleSignInSuccess()` function to handle sign in.
+
+   ```
+	 func handleSignInSuccess(userData: UserData) {
+		 print(userData.profile?.name as Any)
+
+		 self.updateUI()
+
+		 // Check if signed in
+		 // checkIfSignedIn()
+	 }
+   ```
+
+7. Add a `signOut()` action to handle signOut
+
+   ```
+    @IBAction func signOut(_ sender: Any) {
+        // Sign user out
+        Blockstack.shared.signOut()
+        self.updateUI()
+    }
```

Comment:
The issue was that this action was never connected to any button, because there was no "Sign Out" button. You were simply changing the title of the Sign In button to Sign Out, but the functionality wasn't rewired. Generally it's not a good idea to repurpose buttons based on state, so I added a new, hidden button for Sign Out that becomes visible when the user is signed out. Thus, an updateUI() call is needed.

Also, the signature for Blockstack.shared.signOut() has changed (no arguments or completion handler). I fixed this as well in the latest commit.

Reply:
I'll give it a go. When I got this a month back there was only a single button.  The label read *Sign In* regardless of whether the user was or was not signed in.

Reply:
Looks like you also added `optionsContainerView` what is that for?

```
     @IBOutlet weak var optionsContainerView: UIView!
    @IBOutlet weak var resetKeychainButton: UIButton!
```